R/xgx_stat_ci.R

```diff
@@ -99,87 +99,150 @@
+        geom = igeom,
+        position = position,
+        show.legend = show.legend,
+        inherit.aes = inherit.aes,
+        params = list(
+          conf_level = conf_level,
+          distribution = distribution,
+          bins = bins,
+          na.rm = na.rm,
+          ...)
       )
-    } else if (distribution == "binomial") {
-      conf_int_out <- data.frame(
-        y = mean(y),
-        ymin = binom::binom.exact(sum(y), length(y),
-                                  conf.level = 0.95)$lower,
-        ymax = binom::binom.exact(sum(y), length(y),
-                                  conf.level = 0.95)$upper)
-    } else {
-      stop("distribution must be either normal, lognormal, or binomial")
     }
-  }
-
-  ret <- list()
-  for (igeom in geom) {
-    temp <- ggplot2::stat_summary(mapping = mapping, data = data,
-                                  geom = igeom, position = position, ...,
-                                  fun.args = list(), na.rm = na.rm,
-                                  show.legend = show.legend,
-                                  inherit.aes = inherit.aes,
-                                  fun.data = function(y) conf_int(y, conf_level,
-                                                                  distribution)
-    )
 
+    # TODO:
+    #   - This would be cleaner with a dictionary
+    #   - Move this to aes_default in ggproto?
     if (igeom == "point") {
-      if (is.null(temp$aes_params$size)) temp$aes_params$size <- 2
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 2
     }
     else if (igeom == "line") {
-      if (is.null(temp$aes_params$size)) temp$aes_params$size <- 1
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 1
     }
     else if (igeom == "errorbar") {
-      if (is.null(temp$aes_params$size)) temp$aes_params$size <- 1
-      if (is.null(temp$geom_params$width)) {
-        temp$geom_params$width <- 0
-      }
+      if (is.null(lay$aes_params$size)) lay$aes_params$size <- 1
+      if (is.null(lay$geom_params$width)) lay$geom_params$width <- 0
     }
     else if (igeom == "ribbon") {
-      if(is.null(temp$aes_params$alpha)) temp$aes_params$alpha <- 0.25
+      if(is.null(lay$aes_params$alpha)) lay$aes_params$alpha <- 0.25
     }
     else if (igeom == "pointrange") {
-      if(is.null(temp$aes_params$size)) temp$aes_params$size <- 1
-      temp$geom$geom_params$fatten <- 2
+      if(is.null(lay$aes_params$size)){
+        lay$aes_params$size <- 1
+        lay$geom$geom_params$fatten <- 2
+      }
     }
-    
 
-    ret[[paste0("geom_", igeom)]] <- temp
+    lays[[paste0("geom_", igeom)]] <- lay  
   }
-
-  return(ret)
+  
+  return(lays)
 }
+
+
+#' Stat ggproto object for creating ggplot layers of binned confidence intervals
+#'
+#' 
+#' \code{StatCI} returns a ggproto object for plotting mean +/- confidence bins
+#' 
+#'
+#' @return ggplot2 ggproto object
+#'
+#' @export
+StatCI <- ggplot2::ggproto("StatCI", ggplot2::Stat,
```

Comment:
I'm actually not sure if this should go in a separate file or not.  I'm a bit confused about object definitions, but something to think about for clear organization for easier package maintenance.

Reply:
We could potentially make a single file for all ggproto objects such that a simple hierarchical format is kept within the package.  It may be useful to keep the layers in this format as well,  in order to make them simple to find.